docs/proposals/20230407-flexible-managed-k8s-endpoints.md

```diff
@@ -0,0 +1,519 @@
+- `<Infra>Cluster` is watching for changes to `<Infra>ControlPlane` and if the "api server endpoint" on the `<Infra>ControlPlane` CRD instance is not empty then:
+  - Map `<Infra>ControlPlane` to `<Infra>Cluster` and queue event
+  - `<Infra>Cluster` reconciler loop gets the `<Infra>ControlPlane` CRD instance and takes the value for "api server endpoint" and populates `ControlPlaneEndpoint` on the `<Infra>Cluster` CRD instance.
+  - (which will then cause the reconciler for `<Infra>ControlPlane` to run... again)
+
+The implementation of the controllers for Managed Kubernetes would be simplified if there was an option to report the ControlPlaneEndpoint via `<Infra>ControlPlane` instead. Below we will outline two new flows that reduce much of the complexity of the above, while allowing Managed Kubernetes providers to represent their services intuitively.
+
+### Two New Flows
+
+#### Flow 1: `<Infra>Cluster` and `<Infra>ControlPlane`, with `ControlPlaneEndpoint` reported via `<Infra>ControlPlane`
+
+We will describe a CRD composition that adheres to the original separation of concerns of the different provider types as documented in the Cluster API documentation, with a different API Server endpoint reporting flow.
+
+As described above, at present the control plane endpoint must be returned via the `ControlPlaneEndpoint` field on the spec of the `<Infra>Cluster` [reference here](https://cluster-api.sigs.k8s.io/developer/providers/cluster-infrastructure.html). This is OK for self-managed clusters, as a load balancer is usually created as part of the reconciliation. But with Managed Kubernetes services the API Server endpoint usually comes from the service directly, which means that the `<Infra>Cluster` has to get the `ControlPlaneEndpoint` from the managed service so that it can be reported back to CAPI. In practice, this results in `<Infra>Cluster` watching the `<Infra>ControlPlane` and the `<Infra>ControlPlane` watching the `<Infra>Cluster`, and without care this can cause event storms in the CAPI management cluster.
+
+This flow would require making changes to CAPI controllers so that there is an option to report the `ControlPlaneEndpoint` via the `<Infra>ControlPlane` as an alternative to coming from the `<Infra>Cluster`.
+
+Using CAPG as an example:
+
+```go
+type GCPManagedControlPlaneSpec struct {
+  // AddonsConfig defines the addons to enable with the GKE cluster.
+  // +optional
+  AddonsConfig *AddonsConfig `json:"addonsConfig,omitempty"`
+
+  // Logging contains the logging configuration for the GKE cluster.
+  // +optional
+  Logging *ControlPlaneLoggingSpec `json:"logging,omitempty"`
+
+  // EnableKubernetesAlpha will indicate the kubernetes alpha features are enabled
+  // +optional
+  EnableKubernetesAlpha bool
+
+  // ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+  // +optional
+  ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
```

Comment:
This seems to be outdated, now that we introduced the ClusterEndpoint CRD

Reply:
I actually think everything after the `## Background work` sub-heading can be removed. With the evolved proposal, all of this text is just too much IMO.

WDYT @fabriziopandini?

Reply:
answered in https://github.com/kubernetes-sigs/cluster-api/pull/8500/files#r1378834165
(ok to drop if that's ok for you)

Reply:
removed